hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledTaskDescriptor.java

```diff
@@ -90,6 +117,79 @@ void setScheduledFuture(ScheduledFuture<?> future) {
         this.scheduledFuture = future;
     }
 
+    void setResult(Object result) {
+        this.result.set(result);
+    }
+
+    Object getResult()
+            throws ExecutionException, InterruptedException {
+
+        return unpeelFutureOrRoot().get();
+    }
+
+    boolean cancel(boolean mayInterrupt)
+            throws ExecutionException, InterruptedException {
+
+        if (!cancelled.compareAndSet(false, true) || scheduledFuture == null) {
+            return false;
+        }
+
+        // Unpeel Executor's Future (in case of Callable, see. DelegatingCallableTaskDecorator)
+        Future delegatedFuture = null;
```

Comment:
I chose to do my approach to avoid a leaky race-condition between cancelling the task first, only to get it re-scheduled then, before canceling the scheduler one. Not a big deal, just easier to reason about. However, you do have a good point in the second part, which I forgot when I did the refactoring and moved the lines around, I will revert, and handle the exception!!